### ISSUE TITLE
Performance tuning of Spark executor

### ISSUE THREAD

**Issue by vanekjar** [seznam/euphoria#12]
This ticket is considered to be a child of the story #11, where more detailed description of the problem can be found.

**Comment by horkyada** [comment]
In the benchmark I can see two problems:

## Number of stages
The pure spark job looks like this:
![spark-default](https://cloud.githubusercontent.com/assets/25367592/23601728/37e75c1a-024d-11e7-8556-3f5f841ea1a7.png)
The euphoria-spark job implements join by `ReduceStateByKey`, which uses sort. Hence the stages look like this:
![euphoria-stages-grab](https://cloud.githubusercontent.com/assets/25367592/23601781/87fff3d8-024d-11e7-8086-5eb75dfb7849.png)
There is one stage more in the euphoria version. I implemented `JoinTranslator` that directly maps euphoria `Join` to the spark's `Join` operator, but then the stages overview are as followed:
![euphoria-custom-join-stages-grab](https://cloud.githubusercontent.com/assets/25367592/23601855/dfcbb2dc-024d-11e7-8dd7-42d60bca9fb0.png)
Even one stage more!

The problem is that Spark knows that the key (`Tuple(end_of_window, query_string)`) is the same for the `ReduceByKey`s as well as for the `Join` and so it is done in one stage. If you put any `MapToPair` between those two operators, spark cannot be sure about the keys and plans to shuffle the data. That's happening in Euphoria where a custom `KeyExtractor` is used and data are mapped from keyed version to non-keyed version too.

## Size of data
The benchmark is a specific case where the computation is not substant and shuffle data are generally small. Euphoria brings overhead in the size of tha data by inner structure. In this case `WindowedKey` structure:
```java
final class KeyedWindow<W extends Window, K> {
  private final W window;
  private final long timestamp;
  private final K key;
}
```
If the window is `TimeWindow` two more timestamps (start, end) are contained. That means two `long`s overhead compared to the basic spark implementation with only window end timestamp used in tupled key.

**Comment by horkyada** [comment]
**_Number of stages_** 

I found a solution for the benchmark. It consisted of the following issues:

- implement `JoinTranslator` to use native spark's `Join`
- preserve key information among spark translators (do not map from `JavaPairRDD` to `PairRDD`)
- Euphoria does not support explicit key information - i.e. using the same key as in previous operator (omit key extraction)
- detect that windowings from both inputs and the output are overlapping so that they don't need to change their keys (i.e. remap keys) accordingly
- `TimeInterval` in `Time` window has to match (equals & hashCode) by `endTime` only to be able to match long window to short window
- unfortunately some data needs to be part of Spark's key as well as value to match the Euphoria API (provide key as a part answer) which means that more data needs to be shuffled

One stage elimination decreased the overall execution time overhead for the benchmark from 41% to 25%. However, some of the issues are probably unsolvable due to Euphoria API design.

**_Size of data_**

For now it is difficult to reduce the size of the general structures, because all fields are used in various contexts. I experimented with reducing number of bytes in timestamps by setting different time fixed point, but it had negligible gain in execution time.
